generator_nodemodules.go

```diff
@@ -0,0 +1,118 @@
+package dalec
+
+import (
+	"path/filepath"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+const (
+	baseDir      = "/node-mods"
+	yarnCacheDir = baseDir + "/yarn-dalec-cache"
+	npmCacheDir  = baseDir + "/npm-dalec-cache"
+)
+
+func (s *Source) isNodeMod() bool {
+	for _, gen := range s.Generate {
+		if gen.NodeMod != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// HasNodeMods returns true if any of the sources in the spec are node modules.
+func (s *Spec) HasNodeMods() bool {
+	for _, src := range s.Sources {
+		if src.isNodeMod() {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *Spec) HasYarnPackageManager() bool {
+	for _, src := range s.Sources {
+		if src.isNodeMod() {
+			for _, gen := range src.Generate {
+				if gen.NodeMod != nil && gen.NodeMod.PackageManager == "yarn" {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func withNodeMod(g *SourceGenerator, srcSt, worker llb.State, opts ...llb.ConstraintsOpt) func(llb.State) llb.State {
+	return func(in llb.State) llb.State {
+		workDir := "/work/src"
+		joinedWorkDir := filepath.Join(workDir, g.Subpath)
+		srcMount := llb.AddMount(workDir, srcSt)
+
+		var installCmd string
+		if g.NodeMod.PackageManager == "yarn" {
+			installCmd = "npm install --cache " + npmCacheDir + " -g yarn; yarn config set yarn-offline-mirror " + yarnCacheDir + "; yarn install"
+		} else if g.NodeMod.PackageManager == "npm" {
+			installCmd = "npm install --cache " + npmCacheDir
+		} else {
+			panic("unsupported package manager: " + g.NodeMod.PackageManager)
```

Comment:
Can we add a validation for this in `load.go`? Another option is to set a default if not specified